cmd/clusterctl/client/multicluster_discovery.go

```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"context"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/client/cluster"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/client/config"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type MultiClusterTree struct {
+	Name                   string
+	Namespace              string
+	InfrastructureProvider *string
+	WorkloadClusters       []MultiClusterTree
+	Kubeconfig             *string
+}
+
+// MultiDiscovery returns a tree representing the workload cluster discovered in the management cluster.
+func MultiDiscovery(kubeconfigPath string) (*MultiClusterTree, error) {
+	cfgFile := ""
+	configClient, err := config.New(cfgFile)
+	if err != nil {
+		return nil, err
+	}
+
+	clusterClient := cluster.New(cluster.Kubeconfig{Path: kubeconfigPath, Context: ""}, configClient)
+
+	name, err := clusterClient.Proxy().CurrentClusterName()
```

Comment:
This is confusing.
The Cluster name you can find in a Kube config file does not have any relation with namespace/name a Cluster API cluster CR has, and this can lead to misunderstanding.
IMO we should avoid forcing two different concepts into the same struct (unless we properly qualify), even if I understand this somehow can be simpler from the UX, but this will be a violation of boundaries given that we are shaping the return struct of a library method on the bases of a UX concern.